src/cargo/ops/cargo_package/vcs/gix.rs

```diff
@@ -0,0 +1,404 @@
+        .collect();
+    let dirty = !dirty_src_files.is_empty();
+    if !dirty || opts.allow_dirty {
+        // Must check whether the repo has no commit firstly; otherwise `revparse_single` would fail on bare commit repo.
+        // Due to lacking the `sha1` field, it's better not record the `GitVcsInfo` for consistency.
+        let maybe_head_id = repo.head()?.try_peel_to_id_in_place()?;
+        Ok(maybe_head_id.map(|id| GitVcsInfo {
+            sha1: id.to_string(),
+            dirty,
+        }))
+    } else {
+        dirty_src_files.sort_unstable();
+        anyhow::bail!(
+            "{} files in the working directory contain changes that were \
+             not yet committed into git:\n\n{}\n\n\
+             to proceed despite this and include the uncommitted changes, pass the `--allow-dirty` flag",
+            dirty_src_files.len(),
+            dirty_src_files.join("\n")
+        )
+    }
+}
+
+/// Helper to collect dirty statuses for a single repo.
+/// `relative_package_root` is `Some` if the root is a sub-directory of the workdir.
+/// Returns the dirty files outside `relative_package_root`.
+#[must_use]
+fn collect_statuses(
+    repo: &gix::Repository,
+    workdir: &Path,
+    relative_package_root: Option<&Path>,
+    dirty_files: &mut Vec<PathBuf>,
+) -> CargoResult<Vec<PathBuf>> {
+    let statuses = repo
+        .status(gix::progress::Discard)?
+        .dirwalk_options(|opts| {
+            opts.emit_untracked(gix::dir::walk::EmissionMode::Matching)
+                // Also pick up ignored files or whole directories
+                // to specifically catch overzealously ignored source files.
+                // Later we will match these dirs by prefix, which is why collapsing
+                // them is desirable here.
+                .emit_ignored(Some(EmissionMode::CollapseDirectory))
+                .emit_tracked(false)
+                .recurse_repositories(false)
+                .symlinks_to_directories_are_ignored_like_directories(true)
+                .emit_empty_directories(false)
+        })
+        .tree_index_track_renames(TrackRenames::Disabled)
+        .index_worktree_submodules(None)
+        .into_iter(None /* pathspec patterns */)
+        .with_context(|| {
+            format!(
+                "failed to begin git status for repo {}",
+                repo.path().display()
+            )
+        })?;
+
+    let mut dirty_files_outside_of_package_root = Vec::new();
+    for status in statuses {
+        let status = status.with_context(|| {
+            format!(
+                "failed to retrieve git status from repo {}",
+                repo.path().display()
+            )
+        })?;
+
+        let rel_path = gix::path::from_bstr(status.location());
+        let path = workdir.join(&rel_path);
+        if relative_package_root.is_some_and(|pkg_root| !rel_path.starts_with(pkg_root)) {
+            dirty_files_outside_of_package_root.push(path);
+            continue;
+        }
```

Comment:
Out of curiosity. If we adopt the same optimization in git2 version, how the performance would look like?

Reply:
> Maybe only some parts of the package is under git? Like src is a git repo.

It seemed that what the code really was testing is if the file exists, and not if it was tracked by Git. Thus I made it an explicit 'file exists' check, which seemed strange as Cargo won't function without such a manifest. Removing the whole block doesn't fail CI either, so it's a bit unclear why it's there. There is still no tool I know that could do forensics and find everything that Git knows about it, that would certainly be interesting.

Reply:
> Out of curiosity. If we adopt the same optimization in git2 version, how the performance would look like?

The 'architecture' changes here could all be applied to the `git2` implementation, with the same effects. Performance problems should be solved then.